orttraining/orttraining/core/optimizer/lora_conv1d_replacement.cc

```diff
@@ -0,0 +1,141 @@
+          conv1d
+
+  2. the output graph is as follows,
+     the number of matmul is equal to attribute "group" of conv1d
+    conv_input   conv1d.group    conv_weight  conv1d.group
+      \          /                   \         /
+       \        /                   Squeeze   /
+        \      /                       \     /
+          Split                        Split
+          /...\                        /...\
+          ||   |_____________Matmul___||___|
+          ||                   /      ||
+          ||_________Matmul___/_______||
+                       /     /
+                      /...../
+                      Concat
+*/
+namespace onnxruntime {
+bool NodeCanBeReplacedByMatmul(const Node& node) {
+  // if node type is Conv, and attr "dilations" is 1, "kernel_shape" is 1, "stride" is 1, then it can be replaced by matmul
+  // kernel_shape is 1 means it is conv1d
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Conv", {1, 11})) {
+    return false;
+  }
+  const auto* dilations = graph_utils::GetNodeAttribute(node, "dilations");
+  const auto* kernel_shape = graph_utils::GetNodeAttribute(node, "kernel_shape");
+  const auto* stride = graph_utils::GetNodeAttribute(node, "strides");
+  if (dilations == nullptr || kernel_shape == nullptr || stride == nullptr) {
+    return false;
+  }
+  if ((dilations->ints_size() && dilations->ints(0) != 1) || (kernel_shape->ints_size() && kernel_shape->ints(0) != 1) || (stride->ints_size() && stride->ints(0) != 1)) {
+    return false;
+  }
+
+  return true;
+}
+
+void Conv1dToMatmul(Graph& graph, Node& conv) {
+  // shape of conv1d input: [batch_size, in_channels, in_length]
+  // shape of conv1d weight:[output_channels, input_channels/group, kernel_shape], kernel_shape is 1
+  // we need to split the input into "group", and squeeze&split the weight, and then do matmul
+  auto node_description = std::string("LoRAConv1dReplacement");
+  auto execution_provider_type = conv.GetExecutionProviderType();
+  // 1. split conv input
+  auto group_attr = graph_utils::GetNodeAttribute(conv, "group");
+  int64_t group_num = 1;  // default group is 1 from ONNX schema
+  if (group_attr != nullptr) {
+    group_num = group_attr->i();
+  }
+  auto conv1d_input = conv.MutableInputDefs()[0];
+  std::vector<onnxruntime::NodeArg*> conv1d_input_splitted_outputs;
+  for (int i = 0; i < group_num; i++) {
+    conv1d_input_splitted_outputs.push_back(&graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("input_split_output"), nullptr));
+  }
+  auto& input_split = graph.AddNode(graph.GenerateNodeName("Split"), "Split", node_description, {conv1d_input}, {conv1d_input_splitted_outputs});
+  input_split.SetExecutionProviderType(execution_provider_type);
+  input_split.AddAttribute("axis", int64_t(1));
+  auto onnx_opset_version = graph.DomainToVersionMap().at(kOnnxDomain);
+  if (onnx_opset_version >= 18) {
+    input_split.AddAttribute("num_outputs", group_num);
+  }
+  // 2. squeeze conv weight
```

Comment:
Add line break for each code block, for example step 1/step 2
```suggestion

  // 2. squeeze conv weight
```